### PR TITLE
feat: Prepare Additional Conformity Test Scenarios (#3886)

### LINKED ISSUE #3886: Prepare Additional Conformity Test Scenarios for Execution API
### Problem

This issue is a continuation of the work done in [#3823](https://github.com/hiero-ledger/hiero-json-rpc-relay/issues/3823), where we addressed the inability to run the full Ethereum Execution APIs test suite due to limitations in ingesting chain.rlp into the Hedera node.

### Solution

To enable broader test coverage and restore skipped scenarios where possible, we propose the following steps:

1. **Extend and improve** the test file `conformityTests.spec.ts`:

   * Refactor it to support the execution of both standard and adapted (packages/server/tests/acceptance/data/conformity/overwrites) test cases.
   * Improve flexibility to handle custom initialization and validation logic.

2. **Add or update tests for the following JSON-RPC methods** to ensure they are covered in the modified suite:

   **Transaction-related methods:**

   * `eth_getTransactionByBlockHashAndIndex`
   * `eth_getTransactionByBlockNumberAndIndex`
   * `eth_getTransactionByHash`
   * `eth_getTransactionCount`
   * `eth_sendRawTransaction`
   * `eth_getTransactionReceipt`

   **Block-related methods:**

   * `eth_getBlockByHash`
   * `eth_getBlockByNumber`
   * `eth_getBlockTransactionCountByHash`
   * `eth_getBlockTransactionCountByNumber`
   * `eth_getBlockReceipts`

   **Execution/state-related methods:**

   * `eth_getCode`
   * `eth_getStorageAt`
   * `eth_getBalance`

   **Chain/environment info:**

   * `eth_blockNumber`
   * `eth_chainId`
   * `eth_feeHistory`

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by mwb-al**
**Description**:
Add additional conformity tests and extend `conformityTests.spec.ts` to cover Ethereum JSON-RPC methods.

**Related issue(s)**:
Fixes #3886 

**Checklist**

* [ ] Documented (Code comments, README, etc.)
* [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/server/tests/acceptance/data/conformity/utils/validations.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: Apache-2.0
+import Ajv from 'ajv';
+import addFormats from 'ajv-formats';
+import { expect } from 'chai';
+
+import { ErrorResponse, JsonRpcResponse, Method, Schema } from './interfaces';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const execApisOpenRpcData = require('../../../../../../../openrpc_exec_apis.json');
```
Comment (simzzz) [inline packages/server/tests/acceptance/data/conformity/utils/validations.ts:9]: Seems like this file doesn't exist yet when the CI is ran and it is causing it to fial, perhaps you can load it lazily here?

File: packages/server/tests/acceptance/data/conformity/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    console.error(error);
+    if (needError) {
+      return {
+        jsonrpc: '2.0',
+        id: request.id,
+        error: {
+          code: -32603,
+          message: error instanceof Error ? error.message : 'An unknown error occurred',
+        },
+      } as JsonRpcResponse;
```
Comment (simzzz) [inline packages/server/tests/acceptance/data/conformity/utils/utils.ts:55]: Why not return the actual error if such exists? e.g.

     if (error.response?.data) {
        return error.response.data;
      }

File: packages/server/tests/acceptance/data/conformity/utils/processors.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  } else {
+    console.log('Validating a success response.');
+    if (schema && wildcards.length === 0) {
+      console.log('Using schema validation.');
+      const valid = isResponseValid(schema, response);
+      console.log(`Schema validation result: ${valid}`);
+      expect(valid).to.be.true;
+      if (response.result) {
+        console.log('Comparing response result with expected result.');
+        expect(response.result).to.be.equal(JSON.parse(content.response).result);
```
Comment (simzzz) [inline packages/server/tests/acceptance/data/conformity/utils/processors.ts:87]: nit: isn't this expect redundant because of the `isResponseValid` call from earlier?

**Comment by konstantinabl** [comment]
This PR includes a large number of file changes and added lines, so it would be helpful to expand the description with more context.

Specifically:

Could you provide a summary of what was changed in the conformityTest.spec file (e.g., what was removed, what was added, and why)?

Also, please clarify why new tests were added—what functionality or edge cases are they covering?

It would be helpful to list the new tests briefly so reviewers can quickly understand their scope and purpose.

**Review by konstantinabl — CHANGES_REQUESTED** [review]

File: packages/server/tests/acceptance/data/conformity/utils/overwrites.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  NONEXISTENT_TX_HASH,
+  RELAY_URL,
+  transaction1559AndBlockHash,
+  transaction2930AndBlockHash,
+} from './constants';
+import { JsonRpcRequest, Transaction } from './interfaces';
+import { legacyTransaction, transaction1559, transaction1559_2930, transaction2930 } from './transactions';
+import { getTransactionCount } from './utils';
+
+export async function updateRequestParams<T = unknown>(
```
[the remark this answers is not in the document]
Comment (mwb-al) [reply packages/server/tests/acceptance/data/conformity/utils/overwrites.ts:49]: Fixed: db1d6f48

File: packages/server/tests/acceptance/data/conformity/utils/processors.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+import { expect } from 'chai';
+
+import { FileContent } from './interfaces';
+import { updateRequestParams } from './overwrites';
+import { sendRequestToRelay } from './utils';
+import { checkResponseFormat, findSchema, isResponseValid } from './validations';
+
+export function splitReqAndRes(content: string) {
+  /**
```
[the remark this answers is not in the document]
Comment (mwb-al) [reply packages/server/tests/acceptance/data/conformity/utils/processors.ts]: Fixed: 213fa2ac

File: packages/server/tests/acceptance/data/conformity/overwrites/eth_getTransactionCount/get-nonce.io [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+// that queries the transaction count for an address (`sendAccountAddress`) that exists on our test node.
+//
+// Note: This is the original test file, modified for our test purposes: https://github.com/ethereum/execution-apis/blob/main/tests/eth_getTransactionCount/get-nonce.io
+// Only the `params[0]` field and the `result` field have been changed:
+// - `params[0]` now points to `sendAccountAddress`
+// - `result` is set to the value returned by `conformityTest` for that address
+// All other fields must remain unchanged to preserve the integrity of the original test case.
+
+>> {"jsonrpc":"2.0","id":1,"method":"eth_getTransactionCount","params":["0xc37f417fA09933335240FCA72DD257BFBdE9C275","latest"]}
+<< {"jsonrpc":"2.0","id":1,"result":"0x7"}
```
[the remark this answers is not in the document]
Comment (mwb-al) [reply packages/server/tests/acceptance/data/conformity/overwrites/eth_getTransactionCount/get-nonce.io:18]: Fixed: a31f265f

**Review by simzzz — APPROVED** [review]

**Comment by mwb-al** [comment]
@acuarica 
Tests 2–5 depend on Test 1, so after the recent changes to the network state in Test 1, Tests 2–5 started failing. We've decided to introduce changes in batches, starting with this PR which focuses only on fixing Test 1.

I've commented out the other dependent tests for now, since using .skip caused Codacy to raise issues about high cyclomatic complexity and excessive method length.

At the moment, Codacy is reporting decreased test coverage as the main issue. Is there a way to either temporarily lower the test coverage threshold, or mark the affected tests as ignored (e.g., in codecov.yml) while we’re in the process of updating the full set of conformity tests? This would let us merge updates step by step.

**Comment by acuarica** [comment]
Note that the codecov workflow is not required. You can still merge this even is the codecov workflow is failing. No need to change anything in the codecov config.

> _[..]_ is reporting decreased test coverage as the main issue.

Aside note, it looks like codecov is not working properly here given the magnitude in the decrease of code coverage. We are taking at look at improving codecov workflow.

File: tsconfig.json [hunk 7]
```diff
@@ -16,6 +16,7 @@
         /* Specify how TypeScript looks up a file from a given module specifier. */
         "outDir": "./dist",
         /* Specify an output folder for all emitted files. */
+        "resolveJsonModule": true,
```
Comment (acuarica) [inline tsconfig.json:19]: nit: do we actually need this here? I see `tsconfig.json` in `server` package has already this config.

**Review by quiet-node — APPROVED** [review]
LGTM great effort!

**Comment by acuarica** [comment]
Given it's not a priority, let's wait until Monday so we can have more eyes on this. So no need to Dismiss review.

**Comment by konstantinabl** [comment]
@mwb-al Why are we creating locally all the files from the execution apis, instead of fetching them as we used to?

**Comment by mwb-al** [comment]
@konstantinabl 
We’re creating the necessary execution-apis files locally to override specific requests for compatibility with Hedera.
Previously, there were two less maintainable approaches: modifying requests by method name using `if` statements, or using filename-based `switch` logic.
Now, overrides are centralized and aligned with the `node_modules/execution-apis/tests` structure, with all request modifications handled cleanly in a single utility file.
This improves readability, maintainability, and ensures only the required files are customized.

**Comment by mwb-al** [comment]
Thanks for the approval!

Just to clarify, I’m not copying all the files, only the ones I want to overwrite.
The [code](https://github.com/mwb-al/hiero-json-rpc-relay/blob/3886_additional-conformity-tests/packages/server/tests/acceptance/conformityTests.spec.ts#L137) checks which version to use:

```ts
const isCustom = fs.existsSync(path.join(overwritesDirectoryPath, directory, file));
const dir = isCustom ? overwritesDirectoryPath : directoryPath;
```

If a file exists in the `overwrites` folder, it uses that one - otherwise it falls back to the original.
